File: python/1489. Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree.py
class Solution:
    def findCriticalAndPseudoCriticalEdges(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        indexes = dict()
        for i in range(len(edges)):
            indexes[tuple(edges[i])] = i
        edges.sort(key=lambda x: x[2])
        min_weight = self.build_tree(n, edges, None, None)
        critical = []
        pseudo_critical = []
        for e in edges:
            skip_cost = self.build_tree(n, edges, e, None)
            if skip_cost > min_weight:
                critical.append(indexes[tuple(e)])
            else:
                take_cost = self.build_tree(n, edges, None, e)
                if take_cost == min_weight:
                    pseudo_critical.append(indexes[tuple(e)])
        return [critical, pseudo_critical]

    def build_tree(self, n: int, edges: List[List[int]], skip, pick: List[int]) -> int:
        uf = UnionFind(n)
        cost = 0
        if pick:
            uf.union(pick[0], pick[1])
            cost += pick[2]
        for e in edges:
            if e == skip or e == pick:
                continue
            if uf.union(e[0], e[1]):
                cost += e[2]
        return cost if uf.count == 1 else float('+inf')

class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.count = n
    def union(self, a, b: int) -> bool:
        pA, pB = self.find(a), self.find(b)
        if pA == pB:
            return False
        self.parent[pA] = pB
        self.count -= 1
        return True
    def find(self, a: int) -> int:
        if self.parent[a] == a:
            return a
        self.parent[a] = self.find(self.parent[a])
        return self.parent[a]

File: java/1489. Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree.java
class Solution {
    public List<List<Integer>> findCriticalAndPseudoCriticalEdges(int n, int[][] edges) {
        Map<int[], Integer> indexes = new HashMap<>();
        for (int i = 0; i < edges.length; i++) {
            indexes.put(edges[i], i);
        }
        Arrays.sort(edges, (a,b) -> a[2] - b[2]);
        int minWeight = getWeight(n, edges, null, null);
        List<Integer> critical = new ArrayList<>();
        List<Integer> pseudoCritical = new ArrayList<>();
        for (int[] e : edges) {
            int skipWeight = getWeight(n, edges, e, null);
            if (skipWeight > minWeight) critical.add(indexes.get(e));
            else {
                int takeWeight = getWeight(n, edges, null, e);
                if (takeWeight == minWeight) {
                    pseudoCritical.add(indexes.get(e));
                }
            }
        }
        return List.of(critical, pseudoCritical);
    }
    private int getWeight(int n, int[][] edges, int[] skip, int[] pick) {
        UnionFind uf = new UnionFind(n);
        int weight = 0;
        if (pick != null) {
            uf.union(pick[0], pick[1]);
            weight += pick[2];
        }
        for (int[] e : edges) {
            if (e != skip && uf.union(e[0], e[1])) weight += e[2];
        }
        return uf.count == 1 ? weight : Integer.MAX_VALUE;
    }
}

class UnionFind {
    int[] parent;
    int count;
    UnionFind(int n) {
        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
        count = n;
    }
    boolean union(int a, int b) {
        int pa = find(a), pb = find(b);
        if (pa == pb) return false;
        parent[pa] = pb;
        count--;
        return true;
    }
    int find(int a) {
        if (parent[a] == a) return a;
        return parent[a] = find(parent[a]);
    }
}
